source/contracts/trading/Trade.sol

```diff
@@ -15,133 +15,105 @@ import 'libraries/CashAutoConverter.sol';
+        return ICreateOrder(controller.lookup("CreateOrder")).createOrder(_tradeData.sender, Order.getOrderTradingTypeFromMakerDirection(_tradeData.direction), _bestAmount, _tradeData.price, _tradeData.market, _tradeData.outcome, _tradeData.betterOrderId, _tradeData.worseOrderId, _tradeData.tradeGroupId, _tradeData.ignoreShares);
     }
 
-    function fillBestOrder(address _sender, Order.TradeDirections _direction, IMarket _market, uint256 _outcome, uint256 _fxpAmount, uint256 _price, bytes32 _tradeGroupId) internal nonReentrant returns (uint256 _bestFxpAmount) {
+    function fillBestOrder(Data _tradeData) internal nonReentrant returns (uint256 _bestAmount) {
         // we need to fill a BID if we want to SELL and we need to fill an ASK if we want to BUY
-        Order.Types _type = Order.getOrderTradingTypeFromFillerDirection(_direction);
+        Order.Types _type = Order.getOrderTradingTypeFromFillerDirection(_tradeData.direction);
         IOrders _orders = IOrders(controller.lookup("Orders"));
-        bytes32 _orderId = _orders.getBestOrderId(_type, _market, _outcome);
-        _bestFxpAmount = _fxpAmount;
-
-        while (_orderId != 0 && _bestFxpAmount > 0 && gasleft() >= getFillOrderMinGasNeeded()) {
-            uint256 _orderPrice = _orders.getPrice(_orderId);
-            // If the price is acceptable relative to the trade type
-            if (_type == Order.Types.Bid ? _orderPrice >= _price : _orderPrice <= _price) {
-                bytes32 _nextOrderId = _orders.getWorseOrderId(_orderId);
-                _orders.setPrice(_market, _outcome, _orderPrice);
-                _bestFxpAmount = IFillOrder(controller.lookup("FillOrder")).fillOrder(_sender, _orderId, _bestFxpAmount, _tradeGroupId);
-                _orderId = _nextOrderId;
-            } else {
-                _orderId = bytes32(0);
-            }
+        bytes32 _orderId = _orders.getBestOrderId(_type, _tradeData.market, _tradeData.outcome);
+        _bestAmount = _tradeData.amount;
+        uint256 _orderPrice = _orders.getPrice(_orderId);
+        // If the price is acceptable relative to the trade type
+        while (_orderId != 0 && _bestAmount > 0 && _tradeData.loopLimit > 0 && isMatch(_orderId, _type, _orderPrice, _tradeData.price)) {
+            bytes32 _nextOrderId = _orders.getWorseOrderId(_orderId);
+            _orders.setPrice(_tradeData.market, _tradeData.outcome, _orderPrice);
+            _bestAmount = IFillOrder(controller.lookup("FillOrder")).fillOrder(_tradeData.sender, _orderId, _bestAmount, _tradeData.tradeGroupId, _tradeData.ignoreShares);
+            _orderId = _nextOrderId;
+            _orderPrice = _orders.getPrice(_orderId);
+            _tradeData.loopLimit -= 1;
         }
-        if (_orderId != 0) {
+        if (_orderId != 0 && isMatch(_orderId, _type, _orderPrice, _tradeData.price)) {
```

Comment:
This seems a little redundant since `isMatch` already returns false if `_orderId` is 0. Was this intentional?

Reply:
Good point, can just be `isMatch`!